client/engine/messageservice/p2p-message-service/service.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/libp2p/go-libp2p/core/peerstore"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/statechannels/go-nitro/client/engine/store/safesync"
+	"github.com/statechannels/go-nitro/crypto"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+const (
+	MESSAGE_ADDRESS      = "/messages/1.0.0"
+	DELIMITER            = '\n'
+	BUFFER_SIZE          = 1_000
+	NUM_CONNECT_ATTEMPTS = 20
+	RETRY_SLEEP_DURATION = 5 * time.Second
+)
+
+// P2PMessageService is a rudimentary message service that uses TCP to send and receive messages
+type P2PMessageService struct {
+	out chan protocols.Message // for sending message to engine
+
+	peers *safesync.Map[peer.ID]
+
+	quit    chan struct{} // quit is used to signal the goroutine to stop
+	me      types.Address
+	key     p2pcrypto.PrivKey
+	p2pHost host.Host
+}
+
+// Id returns the libp2p peer ID of the message service
+func (ms *P2PMessageService) Id() peer.ID {
+	id, _ := peer.IDFromPrivateKey(ms.key)
+	return id
+}
+
+// AddPeers adds the peers to the message service
+// We ignore peers that are ourselves
+func (ms *P2PMessageService) AddPeers(peers []PeerInfo) {
+
+	for _, p := range peers {
+		// Ignore ourselves
+		if p.Address == ms.me {
+			continue
+		}
+		multi, _ := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", p.IpAddress, p.Port, p.Id))
+
+		// Extract the peer ID from the multiaddr.
+		info, _ := peer.AddrInfoFromP2pAddr(multi)
+		// Add the destination's peer multiaddress in the peerstore.
+		// This will be used during connection and stream creation by libp2p.
+		ms.p2pHost.Peerstore().AddAddrs(info.ID, info.Addrs, peerstore.PermanentAddrTTL)
+
+		ms.peers.Store(p.Address.String(), info.ID)
+	}
+}
+
+// NewMessageService returns a running P2PMessageService listening on the given ip and port
+func NewMessageService(ip string, port int, pk []byte) *P2PMessageService {
+	// We generate a random key using the hash of the pk as a seed
+	// This should mean that the message key is deterministic
+	// TODO: Ideally we would use the pk directly, but I haven't figured out of this is possible with lib p2p
+	hash := sha256.Sum256(pk)
+	seed := big.NewInt(0).SetBytes(hash[:]).Int64()
+	messageKey, _, err := p2pcrypto.GenerateECDSAKeyPair(rand.New(rand.NewSource(seed)))
+	if err != nil {
+		panic(err)
+	}
+	options := []libp2p.Option{libp2p.Identity(messageKey),
+		libp2p.ListenAddrStrings(fmt.Sprintf("/ip4/%s/tcp/%d", ip, port)),
+		libp2p.DefaultTransports,
+		libp2p.NoSecurity,
+		libp2p.DefaultMuxers,
+	}
+	host, err := libp2p.New(options...)
+	if err != nil {
+		panic(err)
+	}
+
+	safePeers := safesync.Map[peer.ID]{}
+	h := &P2PMessageService{
+		out:     make(chan protocols.Message, BUFFER_SIZE),
+		peers:   &safePeers,
+		p2pHost: host,
+		quit:    make(chan struct{}),
+		key:     messageKey,
+		me:      crypto.GetAddressFromSecretKeyBytes(pk),
+	}
+
+	h.p2pHost.SetStreamHandler(MESSAGE_ADDRESS, func(stream network.Stream) {
```

Comment:
If I'm reading this correctly
- each go-nitro protocol message will initiate a new stream
- the message service here reads from the stream until encountering a `\n`ewline
- the read message is parsed & returned to the engine

Are there potential performance impacts here? If Alice and Bob are interacting in a VC, do we expect much performance gain with something like:

``` go
for raw, err := reader.ReadString(DELIMITER); err == nil; {
	m, err := protocols.DeserializeMessage(raw)

	h.checkError(err)
	h.out <- m
}
```

which reads sequential messages on the same stream?

(I've seen a little lib-p2p documentation that refers separately to `streams` and `connections`, so maybe the stream creation / destruction is cheap, given an existing connection?)

Reply:
> which reads sequential messages on the same stream?

Originally the message service would open a `stream` per peer at the start and then reuse that `stream` whenever we want to send messages.

Unfortunately I kept running into `stream reset` errors. It seemed like clients struggled to keep all the streams open (maybe when under load) for the duration of the test. To avoid this I just switched to creating the stream each time we want to send a message.

>(I've seen a little lib-p2p documentation that refers separately to streams and connections, so maybe the stream creation / destruction is cheap, given an existing connection?)

I believe `connections` are seperate than `stream`s. ([Docs](https://github.com/libp2p/specs/blob/master/connections/README.md)). A connection is automatically established when we open our first stream and we then re-use that connection to open streams (I think?). I'm not sure how "cheap" a stream really is.

Longer term maybe our message service could reuse streams but gracefully handle reset errors when they occur?

I should also look at benchmarking the reuse vs new stream scenarios.